blockchain/vm/call_tracer.go

```diff
@@ -0,0 +1,210 @@
+// 1. no error: {output}
+// 2. non-revert error: {to: nil if CREATE, error}
+// 3. revert error without message: {to: nil if CREATE, output, error, reverted{contract}}
+// 4. revert error with message: {to: nil if CREATE, output, error, reverted{contract, message}, revertReason}
+func (c *CallFrame) processOutput(output []byte, err error) {
+	// 1: return output
+	if err == nil {
+		c.Output = common.CopyBytes(output)
+		return
+	}
+
+	// 2,3,4: to = nil if CREATE failed
+	if c.Type == CREATE || c.Type == CREATE2 {
+		c.To = nil
+	}
+
+	// 2: do not return output
+	if !errors.Is(err, ErrExecutionReverted) { // non-revert error
+		c.Error = err.Error()
+		return
+	}
+
+	// 3,4: return output and revert info
+	c.Output = common.CopyBytes(output)
+	c.Error = "execution reverted"
+	c.Reverted = &RevertedInfo{Contract: c.To} // 'To' was recorded when entering this call frame
+
+	// 4: attach revert reason
+	if reason, unpackErr := abi.UnpackRevert(output); unpackErr == nil {
+		c.RevertReason = reason
+		c.Reverted.Message = reason
+	}
+}
+
+// Implements vm.Tracer interface
+type CallTracer struct {
+	callstack       []CallFrame
+	gasLimit        uint64 // saved tx.gasLimit
+	interrupt       atomic.Bool
+	interruptReason error
+}
+
+func NewCallTracer() *CallTracer {
+	return &CallTracer{
+		callstack: make([]CallFrame, 1), // empty top-level frame
+	}
+}
+
+// Transaction start
+func (t *CallTracer) CaptureTxStart(gasLimit uint64) {
+	t.gasLimit = gasLimit
+}
+
+// Transaction end
+func (t *CallTracer) CaptureTxEnd(gasLeft uint64) {
+	t.callstack[0].GasUsed = t.callstack[0].Gas - gasLeft
+}
+
+// Enter top-level call frame
+func (t *CallTracer) CaptureStart(env *EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	toCopy := to
+	t.callstack[0] = CallFrame{
+		Type:  CALL,
+		From:  from,
+		To:    &toCopy,
+		Input: common.CopyBytes(input),
+		Gas:   t.gasLimit, // ignore 'gas' supplied from EVM. Use tx.gasLimit that includes intrinsic gas.
+		Value: value,
+	}
+	if create {
+		t.callstack[0].Type = CREATE
+	}
+}
+
+// Exit top-level call frame
+func (t *CallTracer) CaptureEnd(output []byte, gasUsed uint64, err error) {
+	// gasUsed will be filled by CaptureTxEnd; just process the output
+	t.callstack[0].processOutput(output, err)
+}
+
+// Enter nested call frame
+func (t *CallTracer) CaptureEnter(typ OpCode, from common.Address, to common.Address, input []byte, gas uint64, value *big.Int) {
+	if t.interrupt.Load() {
+		return
+	}
+
+	toCopy := to
+	call := CallFrame{
+		Type:  typ,
+		From:  from,
+		Gas:   gas,
+		To:    &toCopy,
+		Value: value,
+		Input: common.CopyBytes(input),
+	}
+	t.callstack = append(t.callstack, call)
+}
+
+// Exit nested call frame
+func (t *CallTracer) CaptureExit(output []byte, gasUsed uint64, err error) {
+	size := len(t.callstack)
+	if size <= 1 { // just in case; should never happen though because CaptureExit is only called when depth > 0
```

Comment:
According to the comment, `CaptureExit is only called when depth > 0`. Isn't it `size < 1` ?